### PR TITLE
Add no-nested-components ESLint rule

### PR DESCRIPTION
**Pull request by eps1lon**
## Summary

Adds a new `no-nested-components` rule to `eslint-plugin-react-hooks` that triggers on nested component declarations e.g.

```jsx
function Component() {
  const Nested = () => <div />;
  // ^^^^^^^^^^^^^^^^^^^^^^^^^^^ 
  // Component "Nested" is declared during render. 
  // You should move this declaration outside of render to ensure this component's state is persisted across re-renders of its parent. 
  // If this is not a component, or not inside a component or hook, rename it to make sure it is not mistaken for a component e.g. `Icon` -> `renderIcon`.

  return <Nested />
}
```

These declarations are problematic since the nested components will not persist state across re-renders of the parent component.

The [tests are copied from `react/no-nested-unstable-components`](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/tests/lib/rules/no-unstable-nested-components.js) but re-ordered since we currently apply the same logic as rules-of-hooks: If it looks like a component and is inside a component or hooks it cannot be nested inside another component.

Unlike `react/no-nested-unstable-components` we currently don't detect nested component definitions in class components nor do we detect nested class component definitions. Both of these scenarios would be nice to cover as well though!

Open questions: 
- [ ] releaseable? should more cases be covered? better package (name)?
- [ ] rule name bikeshedding
- [ ] internal: inline helpers from RulesOfHooks or created shared helpers?
- [ ] How do we communicate forking of lint rules? What happens in code bases with both `react/no-unstable-nested-components` and `react-hooks/no-nested-components`?

## How did you test this change?

- [x] CI
- [x] [mui/mui](https://github.com/mui/material-ui/)
- [x] Klarna monorepo
- [x] [@AriPerkkio ran it on 20k OSS repos](https://github.com/facebook/react/pull/25360#pullrequestreview-1127521790) One false-positive could be caught before release. The other patterns are consistent with rules-of-hooks

## Why not recommend `react/no-nested-unstable-components`

- [Questionable if this rule will be part of the recommended ruleset in a timely manner.](https://github.com/jsx-eslint/eslint-plugin-react/pull/3444)
- Consistent detection of what's considered a component (e.g. we can detect nested forwardRef or memo component declarations )
- `no-unstable-nested-components` is not actually what we want. It's just about nesting components. Memoizing them will still cause breakage since their state won't be re-useable (remember that useMemo/useCallback are perf optimizations not semantic guarantees unlike `key` and component types)

###  Not detected in proposed rule but detected in `react/no-unstable-nested-components`

A. `<Component footer={() => <div />} />`

This is probably correct to no longer detect since `const footer = () => <div />} />` would also not be considered a component. However, `<Component Footer={() => <div />} />` would also not be detected. This would probably be nice to get into Rules of Hooks regardless.

### Detected in proposed rule but not detected in `react/no-unstable-nested-components`

A.
```jsx
function useHook() {
   return function Component() {}
}
```
Intended. Hooks are called during render which means the component declaration happens during render.

B.
```jsx
function Component() {
   const ConditionalWrapper = ({ condition, wrapper, children }: any) =>
      condition ? wrapper(children) : children
}
```
Intended. Rules of Hooks would allow calling a Hook in `ConditionalWrapper` so we should consider it a nested declaration.

C.

```jsx
function Component() {
  const components = {
    ActionIcon: () => null,
  }
}
```

Intended. Even though flagging this particular example seems silly, Rules of Hooks would allow calling a Hook in `ActionIcon` so we should consider it a nested declaration.

### CODE REVIEW

**Review by AriPerkkio — COMMENTED** [review]
I ran this through 2000 repositories. No crashes but some false positives. Most are from uppercase functions which do not return JSX - the `eslint-plugin-react/no-unstable-nested-components` rule especially checks whether function is returning JSX or `React.createElement`. Maybe this one should do too.

Copy-paste ready test cases:

```jsx
{
  code: normalizeIndent`
    function ParentComponent() {
      function NotComponent() {
        console.log('Not returning JSX');
      }

      return <button onClick={NotComponent} />;
    }
  `,
},
{
  code: normalizeIndent`
    function ParentComponent() {
      const inputKeyDownHandlers = useMemo(() => ({
        ArrowDown(event) {
          event.preventDefault();
          // Do something
        },
        ArrowUp(event) {
          event.preventDefault();
          // Do something
        },
      }));
  
      return <ComplexComponent handlers={inputKeyDownHandlers} />;
    }
  `,
},
{
  code: normalizeIndent`
    const MyComponent = React.memo(
      React.forwardRef((props, ref) => {
        return <div />
      })
    )
  `,
},
```

<details>
  <summary>Uppercase component not returning JSX</summary>

## Rule: react-hooks/no-nested-components

-   Message: `Component "CleanupInput" is declared during render. You should move this declaration outside of render to ensure this component's state is persisted across re-renders of its parent. If this component is memoized, you should still refactor the component to be able to move it outside of render. If you want to reset its state use a key instead (see https://reactjs.org/docs/lists-and-keys.html#keys).`
-   Path: `axmz/react-searchbox-awesome/src/components/Search.jsx`
-   [Link](https://github.com/axmz/react-searchbox-awesome/blob/HEAD/src/components/Search.jsx#L84-L88)

```jsx
  82 |
  83 |   // Cleanup input
> 84 |   function CleanupInput() {
     |   ^^^^^^^^^^^^^^^^^^^^^^^^^
> 85 |     inputRef.current.value = ""
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 86 |     inputRef.current.dispatchEvent(inputEvent)
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 87 |     inputRef.current.focus();
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 88 |   }
     | ^^^^
  89 |
  90 |   // Click handler
  91 |   const clickHandler = (e) => {
```

</details>

<details>
    <summary>Memo + forwardRef</summary>

## Rule: react-hooks/no-nested-components

-   Message: `Component "Unknown" is declared during render. You should move this declaration outside of render to ensure this component's state is persisted across re-renders of its parent. If this component is memoized, you should still refactor the component to be able to move it outside of render. If you want to reset its state use a key instead (see https://reactjs.org/docs/lists-and-keys.html#keys).`
-   Path: `alloc/wana/packages/babel-plugin-add-react-displayname/spec/fixtures/multipleWraps.tsx`
-   [Link](https://github.com/alloc/wana/blob/HEAD/packages/babel-plugin-add-react-displayname/spec/fixtures/multipleWraps.tsx#L2-L4)

```tsx
  1 | export const Test1 = React.memo(
> 2 |   React.forwardRef((props, ref) => {
    |                    ^^^^^^^^^^^^^^^^^
> 3 |     return <div />
    | ^^^^^^^^^^^^^^^^^^
> 4 |   })
    | ^^^^
  5 | )
  6 |
  7 | const Test2 = React.memo(
```

</details>


<details>
    <summary>Hook-like, not returninig JSX</summary>

## Rule: react-hooks/no-nested-components

-   Message: `Component "useBlock" is declared during render. You should move this declaration outside of render to ensure this component's state is persisted across re-renders of its parent. If this component is memoized, you should still refactor the component to be able to move it outside of render. If you want to reset its state use a key instead (see https://reactjs.org/docs/lists-and-keys.html#keys).`
-   Path: `BuilderIO/jsx-lite/packages/core/src/__tests__/data/blocks/builder-render-block.raw.tsx`
-   [Link](https://github.com/BuilderIO/jsx-lite/blob/HEAD/packages/core/src/__tests__/data/blocks/builder-render-block.raw.tsx#L48-L54)

```tsx
  46 |       return getBlockProperties(state.useBlock);
  47 |     },
> 48 |     get useBlock() {
     |                 ^^^^
> 49 |       return getProcessedBlock({
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 50 |         block: props.block,
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 51 |         state: builderContext.state,
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 52 |         context: builderContext.context,
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 53 |       });
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 54 |     },
     | ^^^^^^
  55 |     get actions() {
  56 |       return getBlockActions({
  57 |         block: state.useBlock,
```
</details>

<details>
    <summary>Memoized objcet with uppercase functions</summary>

## Rule: react-hooks/no-nested-components

-   Message: `Component "ArrowDown" is declared during render. You should move this declaration outside of render to ensure this component's state is persisted across re-renders of its parent. If this component is memoized, you should still refactor the component to be able to move it outside of render. If you want to reset its state use a key instead (see https://reactjs.org/docs/lists-and-keys.html#keys).`
-   Path: `downshift-js/downshift/src/hooks/useCombobox/index.js`
-   [Link](https://github.com/downshift-js/downshift/blob/HEAD/src/hooks/useCombobox/index.js#L155-L162)

```js
  153 |   const inputKeyDownHandlers = useMemo(
  154 |     () => ({
> 155 |       ArrowDown(event) {
      |                ^^^^^^^^^
> 156 |         event.preventDefault()
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 157 |         dispatch({
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 158 |           type: stateChangeTypes.InputKeyDownArrowDown,
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 159 |           shiftKey: event.shiftKey,
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 160 |           getItemNodeFromIndex,
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 161 |         })
      | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 162 |       },
      | ^^^^^^^^
  163 |       ArrowUp(event) {
  164 |         event.preventDefault()
  165 |         dispatch({
```

</details>

**Comment by eps1lon** [comment]
>  Memoized objcet with uppercase functions + Uppercase component not returning JSX

In these false-positives, what would the existing react-hooks rules say? The idea was that it should flag everything that could also contain a hooks call. So "false-positives" of this rule would be false-negatives of the other rules.

> Memo + forwardRef

The memo+forwardRef is definitely something we should not flag 👍🏻 

> Hook-like, not returninig JSX

That case needs a better message but is not a false-positive. The function is considered a hook and can thus be called during render.

**Comment by AriPerkkio** [comment]
> > Memoized object with uppercase functions + Uppercase component not returning JSX
>
> In these false-positives, what would the existing react-hooks rules say?

Existing `eslint-plugin-react-hooks` rules won't report anything in these cases. And the existing `eslint-plugin-react/no-unstable-nested-components` doesn't either since functions are not returning JSX or output of `React.createElement`.

> In those cases the author should just rename `ArrowDown` to `arrowDown`.

I agree this should be the recommended solution but not sure how this could be informed to users. Now this rule would confuse users by calling these functions components. Instead the rule should detect that these are not returning JSX, and recommend users to rename the **function** into lowercase.

> > I ran this through 2000 repositories.
> 
> How did you do that?

Started your react fork in Github Codespaces and installed [`eslint-remote-tester`](https://github.com/AriPerkkio/eslint-remote-tester#developing-new-eslint-rules). I'm always using it to see how rules under development work in real life projects. Unit tests are simply not enough.

<details>
  <summary>eslint-remote-tester.config.js</summary>

```js
const eslintRemoteTesterRepositories = require('eslint-remote-tester-repositories');

module.exports = {
  repositories: eslintRemoteTesterRepositories
    .getRepositories({ randomize: false })
    .slice(0, 2000),

  pathIgnorePattern: eslintRemoteTesterRepositories.getPathIgnorePattern(),

  extensions: ['.js', '.jsx', '.ts', '.tsx'],

  rulesUnderTesting: ['react-hooks/no-nested-components'],

  cache: false,

  eslintrc: {
    root: true,
    env: {
      es6: true,
    },
    parser: '@typescript-eslint/parser',

    parserOptions: {
      ecmaVersion: 2020,
      sourceType: 'module',
      ecmaFeatures: {
        jsx: true,
      },
    },
    plugins: ['react-hooks'],
    rules: {
      'react-hooks/no-nested-components': ['error'],
    },
  },
};
```
</details>

**Comment by eps1lon** [comment]
> Existing eslint-plugin-react-hooks rules won't report anything in these cases. 

Which is the problem. React thinks these are function components and allows calling hooks. And if it's a component, it should not allow nested declarations. Either we start flagging hooks calls in these cases or we continue flagging nested component declarations. But having an inconsistent detection is the worst scenario since it requires knowledge how individual **rules** detect components. When you should just need to know what the **plugin** (or rather React itself) considers a function component.

> I agree this should be the recommended solution but not sure how this could be informed to users. Now this rule would confuse users by calling these functions components.

This has been the narrative since rules-of-hooks was released. We can amend the error message to explain why this is considered a component and how to fix it by renaming.

>  Instead the rule should detect that these are not returning JSX

A component does not need to return JSX. It can return primitives as well. I would not change this detection mechanism in this PR. It should be consistent with the existing rules-of-hooks. If there are real-world scenarios where renaming does not make sense, we can follow-up later.

File: packages/eslint-plugin-react-hooks/README.md [hunk 1]
```diff
@@ -70,6 +71,93 @@ We suggest to use this option **very sparingly, if at all**. Generally saying, w
...
+  function UnstableNestedComponent() {
+    return <div />;
+  }
+
+  return (
+    <div>
+      <UnstableNestedComponent />
+    </div>
+  );
+}
```
[the remark this answers is not in the document]
Comment (eps1lon) [reply packages/eslint-plugin-react-hooks/README.md:98]: Good point. Just copied these tests but I should make a pass regarding naming

**Pull request closed without merging**